clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -483,6 +483,7 @@ public void logUnused() {
                 resolvedOriginals.putAll(result.data());
             }
         }
+        providers.values().forEach(x -> Utils.closeQuietly(x, "config provider"));
```

Comment:
if an exception occurs between line 477 and line 486, then the close function won't be called.  The solution would be to either create a type for `providers` that extends `Map<String, ConfigProvider>` and implements Autoclosable, or to simply put an explicit `try`/`finally` block here to ensure that the close function is called in every case.  That also implies that `instantiateConfigProviders` should be modified so that if an exception is thrown from inside it, any previously opened `ConfigProvider` instances are closed.

Reply:
Good observations @jherico . I believe the latter approach, of closing all the instantiated providers if an exception occurs in both cases, would be the most straightforward fix. 

If you'd be interested in submitting a fix, that would be very welcome!